### ISSUE TITLE
析构函数里面一个点

### ISSUE THREAD

**Issue by Staok** [lzpong/threadpool#40]
```
	inline ~threadpool()
	{
		_run=false;
		_task_cv.notify_all(); // 唤醒所有线程执行
		for (thread& thread : _pool) {
			//thread.detach(); // 让线程“自生自灭”
			if (thread.joinable())
				thread.join(); // 等待任务结束， 前提：线程一定会执行完
		}
	}
```

`_run=false;` 这句话需要 条件变量的锁包着，参考原作者的代码：
https://github.com/progschj/ThreadPool/blob/master/ThreadPool.h#L90

另外，`// thread.detach();` 这句话虽然注释了，但是可以去掉，以防其他人解掉注释，给线程池添加了一些死循环的函数，这样线程池里面的线程永远不能释放了，造成长期运行的资源浪费、资源挤占。原作者这里只用 join 是有原因的。

另外，类里面的变量不必加下划线前缀，规范使用是用 m 前缀即可，比如 _run 改为 mRun。

**Comment by lx7741** [comment]
有时候析构会卡死。
改成
_run=false;
{
          lock_guard<mutex> lock{ _lock };
}

**Comment by hide955** [comment]
> 有时候析构会卡死。 改成 _run=false; { lock_guard lock{ _lock }; }
{
lock_guard lock{ _lock };
_run=false;
}
我看到的都是这样. 按照你的说法. 什么情况下会卡死呢?

**Comment by lzpong** [comment]
[Enhance thread safety in threadpool destructor and commit methods](https://github.com/lzpong/threadpool/commit/031fe670f042045cee43539a54abf2229bd017f9)

**Comment by Staok** [comment]
done.
多用最先进 AI 反复找找问题、多写测试。
